parsers/jats/PublicationParser.inc.php

```diff
@@ -121,12 +121,144 @@ public function getPublication(): \Publication
+				$this->_createDependentFile($genreId, $submission, $insertedSubmissionFile, $userId, $fileType, $fileName, \SUBMISSION_FILE_DEPENDENT, \ASSOC_TYPE_SUBMISSION_FILE, false, $insertedSubmissionFile->getFileId(), false, $dependentFilePath);
+			}
+		}
+
+		foreach ($this->select('//graphic') as $asset) {
+			$assetFilename = $asset->getAttribute('xlink:href');
+			$dependentFilePath = dirname($filename) . DIRECTORY_SEPARATOR . $assetFilename;
+			if (file_exists($dependentFilePath)) {
+				$fileType = pathinfo($dependentFilePath, PATHINFO_EXTENSION);
+
+				$genreId = $this->_getGenreId($this->getContextId(), $fileType);
+				$this->_createDependentFile($genreId, $submission, $insertedSubmissionFile, $userId, $fileType, $assetFilename, \SUBMISSION_FILE_DEPENDENT, \ASSOC_TYPE_SUBMISSION_FILE, false, $insertedSubmissionFile->getFileId(), false, $dependentFilePath);
+			}
+		}
+
+	}
+
+
+	/**
+	 * Creates a dependent file
+	 *
+	 * @param $genreId  int
+	 * @param $submission Submission
+	 * @param $submissionFile SubmissionFile
+	 * @param $userId int
+	 * @param $fileType string
+	 * @param $fileName string
+	 * @param bool $fileStage
+	 * @param bool $assocType
+	 * @param bool $sourceRevision
+	 * @param bool $assocId
+	 * @param bool $sourceFileId
+	 * @param $filePath string
+	 * @return void
+	 */
+	protected function _createDependentFile($genreId, $submission, $submissionFile, $userId, $fileType, $fileName, $fileStage = false, $assocType = false, $sourceRevision = false, $assocId = false, $sourceFileId = false, $filePath) {
+
+		$fileSize = filesize($filePath);
+
+		$submissionFileDao = \DAORegistry::getDAO('SubmissionFileDAO');
+		$newFile = $submissionFileDao->newDataObjectByGenreId($genreId);
+		$newFile->setSubmissionId($submission->getId());
+		$newFile->setSubmissionLocale($submission->getLocale());
+		$newFile->setGenreId($genreId);
+		$newFile->setFileStage($fileStage);
+		$newFile->setDateUploaded(\Core::getCurrentDate());
+		$newFile->setDateModified(\Core::getCurrentDate());
+		$newFile->setUploaderUserId($userId);
+		$newFile->setFileSize($fileSize);
+		$newFile->setFileType($fileType);
+
+		if (isset($fileName)) $newFile->setOriginalFileName($fileName);
+		if (isset($fileName)) $newFile->setName($fileName, $submission->getLocale());
+		if (isset($assocType)) $newFile->setAssocType($assocType);
+		if (isset($sourceRevision)) $newFile->setSourceRevision($sourceRevision);
+		if (isset($assocId)) $newFile->setAssocId($assocId);
+		if (isset($sourceFileId)) $newFile->setSourceFileId($sourceFileId);
+
+		$insertedMediaFile = $submissionFileDao->insertObject($newFile, $filePath);
+
+	}
+
+	/**
+	 * @param $contextId
+	 * @param $extension
+	 * @return int
+	 */
+	private function _getGenreId($contextId, $extension) {
+
+		$genreDao = \DAORegistry::getDAO('GenreDAO');
+		$genres = $genreDao->getByDependenceAndContextId(true, $contextId);
+
+		while ($candidateGenre = $genres->next()) {
```

Comment:
I think you can cache the genres in a static array (I've used local static variables in some places 👀), keyed by the contextId, to avoid hitting the database.

In general I think such cache should exist natively in the application (and be configurable in case the mechanism to trigger the cache invalidation is complex), I'll try to propose something later o_O

Reply:
Rewritten to cache, simplify, and move to `BaseParser`.